core/common/src/main/java/alluxio/AbstractThriftClient.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio;
+
+import alluxio.exception.AlluxioException;
+import alluxio.thrift.AlluxioService;
+import alluxio.thrift.AlluxioTException;
+import alluxio.thrift.ThriftIOException;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * The base class for clients that use {@link alluxio.network.connection.ThriftClientPool}.
+ */
+// TODO(peis): Rename this to Client once we have deprecated AbstractClient and Client class.
+@ThreadSafe
+public abstract class AbstractThriftClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final int RPC_MAX_NUM_RETRY = 30;
+
+  /**
+   * @return a Thrift service client
+   */
+  protected abstract AlluxioService.Client acquireClient() throws IOException;
+
+  /**
+   * @param client the client to release
+   */
+  protected abstract void releaseClient(AlluxioService.Client client);
+
+  /**
+   * The RPC to be executed in {@link #retryRPC(RpcCallable)}.
+   *
+   * @param <V> the return value of {@link #call(AlluxioService.Client)}
+   */
+  protected interface RpcCallable<V> {
+    /**
+     * The task where RPC happens.
+     *
+     * @return RPC result
+     * @throws TException when any exception defined in thrift happens
+     */
+    V call(AlluxioService.Client client) throws TException;
+  }
+
+  /**
+   * Same with {@link RpcCallable} except that this RPC call throws {@link AlluxioTException} and
+   * is to be executed in {@link #retryRPC(RpcCallableThrowsAlluxioTException)}.
+   *
+   * @param <V> the return value of {@link #call(AlluxioService.Client)}
+   */
+  protected interface RpcCallableThrowsAlluxioTException<V> {
+    /**
+     * The task where RPC happens.
+     *
+     * @return RPC result
+     * @throws AlluxioTException when any {@link AlluxioException} happens during RPC and is wrapped
+     *         into {@link AlluxioTException}
+     * @throws TException when any exception defined in thrift happens
+     */
+    V call(AlluxioService.Client client) throws AlluxioTException, TException;
```

Comment:
Because I want to release the client in AbstractThriftClient. If I don't pass, I need to add a try-finally block for every rpc call. I need to do this:

retryRPC(new RpcCallableThrowsAlluxioTException<Void>() {
      @Override
      public Void call() throws AlluxioTException, TException {
       try {
       AlluxioService.Client client = acquireClient();
        if (client instanceof BlockWorkerClientService.Client) {
          ((BlockWorkerClientService.Client) client).cancelBlock(mSessionId, blockId);
          return null;
        } else {
          throw new RuntimeException(
              "Unexpected type. BlockWorkerClientService.Client is expected.");
        }
      } finally {
        releaseClient(client);
     }


I am fine with this if you prefer.